src/cs-graphics/GlareMipMap.cpp

```diff
@@ -0,0 +1,448 @@
+  // details.
+
+  glUseProgram(mComputeProgram);
+
+  glBindImageTexture(0, hdrBufferComposite->GetId(), 0, GL_FALSE, 0, GL_READ_ONLY, GL_RGBA32F);
+
+  // The asymmetric variant requires the projection and the inverse projection matrices.
+  if (glareMode == HDRBuffer::GlareMode::eAsymmetricGauss) {
+    std::array<GLfloat, 16> glMatP{};
+    glGetFloatv(GL_PROJECTION_MATRIX, glMatP.data());
+    glm::mat4 matP    = glm::make_mat4x4(glMatP.data());
+    glm::mat4 matInvP = glm::inverse(matP);
+    glUniformMatrix4fv(mUniforms.projectionMatrix, 1, GL_FALSE, glm::value_ptr(matP));
+    glUniformMatrix4fv(mUniforms.inverseProjectionMatrix, 1, GL_FALSE, glm::value_ptr(matInvP));
+  }
+
+  for (int level(0); level < mMaxLevels; ++level) {
+    glUniform1i(mUniforms.level, level);
+
+    for (int pass(0); pass < 2; ++pass) {
+      VistaTexture* input       = this;
+      VistaTexture* output      = this;
+      int           inputLevel  = level;
+      int           outputLevel = level;
+
+      // level  pass   input   inputLevel output outputLevel     blur      samplesHigherLevel
+      //   0     0   hdrbuffer    0        temp      0        horizontal          true
+      //   0     1     temp       0        this      0         vertical           false
+      //   1     0     this       0        temp      1        horizontal          true
+      //   1     1     temp       1        this      1         vertical           false
+      //   2     0     this       1        temp      2        horizontal          true
+      //   2     1     temp       2        this      2         vertical           false
+      //   3     0     this       2        temp      3        horizontal          true
+      //   3     1     temp       3        this      3         vertical           false
+
+      if (pass == 0) {
+        output     = mTemporaryTarget;
+        inputLevel = std::max(0, inputLevel - 1);
+      }
+
+      if (pass == 1) {
+        input = mTemporaryTarget;
+      }
+
+      glUniform1i(mUniforms.pass, pass);
+
+      int width = static_cast<int>(
+          std::max(1.0, std::floor(static_cast<double>(static_cast<int>(mHDRBufferWidth / 2)) /
+                                   std::pow(2, level))));
+      int height = static_cast<int>(
+          std::max(1.0, std::floor(static_cast<double>(static_cast<int>(mHDRBufferHeight / 2)) /
+                                   std::pow(2, level))));
+
+      glBindImageTexture(1, input->GetId(), inputLevel, GL_FALSE, 0, GL_READ_ONLY, GL_RGBA32F);
+      glBindImageTexture(2, output->GetId(), outputLevel, GL_FALSE, 0, GL_WRITE_ONLY, GL_RGBA32F);
+
+      glMemoryBarrier(GL_SHADER_IMAGE_ACCESS_BARRIER_BIT);
+      glDispatchCompute(static_cast<uint32_t>(std::ceil(1.0 * width / 16)),
+          static_cast<uint32_t>(std::ceil(1.0 * height / 16)), 1);
+    }
+  }
+
+  glBindImageTexture(0, 0, 0, GL_FALSE, 0, GL_WRITE_ONLY, GL_RGBA32F);
+  glBindImageTexture(1, 0, 0, GL_FALSE, 0, GL_READ_ONLY, GL_RGBA32F);
+}
```

Comment:
Do we also need to unbind the texture ate location 2 or is it used immediately after?

Reply:
Same as above :smile: